File: app/src/main/java/com/pr0gramm/app/ui/TagInputView.java
package com.pr0gramm.app.ui;

import android.annotation.SuppressLint;
import android.content.Context;
import android.widget.ArrayAdapter;
import android.widget.MultiAutoCompleteTextView;

import com.google.common.base.Charsets;
import com.google.common.collect.ImmutableList;
import com.google.common.io.CharStreams;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.util.Collections;
import java.util.List;

/**
 */
public class TagInputView {
    private static final Logger logger = LoggerFactory.getLogger(TagInputView.class);

    private TagInputView() {
    }

    /**
     * Parses the list of tags. It is provided as one tag per line in utf8 encoding.
     */
    @SuppressLint("NewApi")
    private static List<String> loadTagJson(Context context) {
        try (InputStream stream = context.getAssets().open("tags.txt")) {
            try (Reader reader = new InputStreamReader(stream, Charsets.UTF_8)) {
                ImmutableList.Builder<String> result = ImmutableList.builder();
                for (String line : CharStreams.readLines(reader)) {
                    line = line.trim();
                    if (line.length() > 1) {
                        result.add(line.trim());
                    }
                }

                return result.build();
            }
        } catch (Exception error) {
            logger.error("Could not load list of tags", error);
            return Collections.emptyList();
        }
    }

    public static void setup(MultiAutoCompleteTextView tagInput) {
        if (TAGS == null) {
            // load the tags from the compressed json file
            TAGS = loadTagJson(tagInput.getContext());
        }

        // get the auto-suggestion list.
        ArrayAdapter<String> adapter = new ArrayAdapter<>(tagInput.getContext(),
                android.R.layout.simple_dropdown_item_1line,
                TAGS);

        tagInput.setTokenizer(new MultiAutoCompleteTextView.CommaTokenizer());
        tagInput.setAdapter(adapter);
    }

    // holds the list of tags
    private static List<String> TAGS;
}

File: app/src/main/java/com/pr0gramm/app/ui/TagInputView.kt
package com.pr0gramm.app.ui

import android.content.Context
import android.widget.ArrayAdapter
import android.widget.MultiAutoCompleteTextView
import com.pr0gramm.app.services.config.Config
import com.pr0gramm.app.util.BackgroundScheduler
import com.pr0gramm.app.util.directKodein
import com.pr0gramm.app.util.time
import com.trello.rxlifecycle.kotlin.bindToLifecycle
import org.kodein.di.erased.instance
import org.slf4j.LoggerFactory
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import java.io.InputStreamReader

/**
 */
object TagInputView {
    private val logger = LoggerFactory.getLogger("TagInputView")

    /**
     * Parses the list of tags. It is provided as one tag per line in utf8 encoding.
     */
    private fun loadTagJson(context: Context): List<String> {
        val config = context.directKodein.instance<Config>()

        val questionableTags = config.questionableTags.map { it.toLowerCase() }

        try {
            return logger.time("Loading tags from file") {
                context.assets.open("tags.txt").use { stream ->
                    InputStreamReader(stream, Charsets.UTF_8).use { reader ->
                        reader.readLines().mapNotNull { line ->
                            line.trim().takeIf { tag ->
                                val lower = tag.toLowerCase()
                                tag.length > 1 && questionableTags.none { lower.contains(it) }
                            }
                        }
                    }
                }
            }
        } catch (error: Exception) {
            logger.error("Could not load list of tags", error)
            return emptyList()
        }
    }

    private lateinit var TAGS: Observable<List<String>>


    fun initialize(context: Context) {
        TAGS = Observable
                .fromCallable { loadTagJson(context.applicationContext) }
                .subscribeOn(BackgroundScheduler.instance())
                .cache()
    }

    fun setup(tagInput: MultiAutoCompleteTextView) {
        TAGS.first().observeOn(AndroidSchedulers.mainThread()).bindToLifecycle(tagInput).subscribe { tags ->
            val adapter = ArrayAdapter(tagInput.context,
                    android.R.layout.simple_dropdown_item_1line,
                    tags)

            tagInput.setTokenizer(MultiAutoCompleteTextView.CommaTokenizer())
            tagInput.setAdapter(adapter)
        }
    }
}
